Add lsp (cquery and ccls) backend support to c-c++ layer

cormacc opened the pull request:
**N.B. This PR should be considered as on hold pending acceptance/rejection of #11351 . The current commit ~~is functional for evaluation purposes, however the keybinding logic will be updated to reflect the navigation style preferences in #11351 if that gets merged.~~ includes the lsp layer changes to facilitate evaluation in the interim**

Migrated the cquery and ccls work from a separate `lsp-c-c++` layer to the existing `c-c++` layer.
Enabled by adding the following to your spacemacs dotfile.

``` elisp
  (setq-default dotspacemacs-configuration-layers
                '(; ... other layers ...
                  (c-c++ :variables c-c++-backend 'lsp-cquery))) ;or  'lsp-ccls
```
More detailed information has been added to the README.

The `clang-format-` keybindings will be added automatically if either lsp backend is selected.

This is just to get the ball rolling -- appears to work as the previous `lsp-c-c++` layer based on some quick and dirty testing with both cquery and ccls backends.

The rough edges I'm aware of so far...

1. ~~Setting `c-c++-lsp-sem-highlight-rainbow` to `t` results in an error. The macros `<cquery/ccls>-use-default-rainbow-sem-highlight` don't seem to be in scope when the layer is loaded, though they're provided by their respective packages -- suggestions welcome. This did work in some earlier incarnation of the `lsp-c-c++` layer, but had stopped doing so before this most recent bit of integration. No idea whether it's something I've done or an upstream change affecting scope of package macros.~~
Fixed based on myrgy's suggestion
2. ~~RTags is enabled by the existing mechanism of setting `c-c++-enable-rtags-support` to `t`. It would be neater to set `c-c++-backend` to `'rtags` instead, but I don't know what the policy is on changing layer configuration mechanisms. Again, guidance welcome. The existing config var may be `nil` `t` or `'no-completion`. If moving to using `c-c++-backend` desirable, would it be preferable to
    (a) allow values of `'rtags` and `'rtags-no-completion`, or
    (b) just `rtags` and add a new `c-c++-rtags-completion` variable with a default value of `t`?~~
RTags now enabled using the new `c-c++-backend` config.el variable. Completion enabled by default, disabled by setting new `c-c-++-enable-rtags-completion` config.el variable to `nil`

cormacc commented:
@myrgy You might take a look when you get a chance?

@sdwolfz I've derived the backend selection/configuration logic from some of the recent examples you suggested (mostly some useful discussion in the python layer PR)

@maskray, @sebastiencs either of you have any enlightenment to offer on the `-use-default-sem-rainbow-highlight` issue? Also any useful backend-specific bindings I've missed?

Compro-Prasad commented:
I think there should be an universal backend variable. This would prevent conflicts in between backends such that the variable accepts only a single symbol.

cormacc commented:
@Compro-Prasad That's my own view too - just want to get confirmation from maintainers that that's acceptable before making the change.

myrgy commented:
@cormacc , Thanks a lot for outstanding work!
Will take a look on Monday.

sdwolfz commented:
Some comments related to variables, docs and formatting. I have not tested it locally yet.

myrgy commented on layers/+lang/c-c++/README.org:
```diff
@@ -52,22 +65,121 @@ by setting the variable =c-c++-default-mode-for-headers= to =c++-mode=.
+  (setq-default dotspacemacs-configuration-layers
+                '(
+                  lsp
+                  (c-c++ :variables c-c++-backend `cquery) ;or 'ccls
```
why do we need to create new layers? we have rtas inside c++ layer, so it's more similar to different backends - rtags, ccls, cquery.
Real issue there is that it's required to enable lsp layers manually. It might be great to use
`(configuration-layer/declare-layer 'base)` but with condition.
@syl20bnr , could you please tell if it's possible?

sdwolfz replied:
No new layers, just the values of the `c-c++-backend` variable to be prefixed with `lsp`.

myrgy replied:
For me it seem to be unnecessary, what I'd really want is that if we'll be unable to enable automatic dependency from lsp layer, some error message should be printed.
It could be done by addiing empty test package, which will check `c-c++-backend` and if lsp is not available.

cormacc replied:
Re. having to add the `lsp` layer explicitly to your dotfile, creating a `layers.el` with a conditional layer declaration works to get around it:

``` elisp
(when (member c-c++-backend '(lsp-cquery lsp-ccls))
  (configuration-layer/declare-layer 'lsp))
```
is this acceptable @sdwolfz ?

I.e. it appears `config.el` is loaded before `layers.el`.
I'd tried previously using the `spacemacs//c-c++-lsp-backend` function defined in `funcs.el`, but `layers.el` appears to be loaded before it.

cormacc replied:
The current commit amendment still has some logic to display warnings if `lsp` layer not loaded, which should be unnecessary if the conditional layer dependency declaration is acceptable. Easier to strip it out in another amendment if I get the green light than to reinstate it if not :)

sdwolfz replied:
I think it's good to have it. I don't see it being used in other layers but as long as it does not caus trouble when using a different backend (rtags) then it's a good addition.

myrgy commented on layers/+lang/c-c++/README.org:
```diff
@@ -52,22 +65,121 @@ by setting the variable =c-c++-default-mode-for-headers= to =c++-mode=.
+
+***** Additional configuration options
+Both supported backends store their index cache in a subdirectory of the project dir by default. This can be overridden by
+specifying an explicit =lsp-c-c++-cache-dir=.
```
At least `ccls` support absolute path.

myrgy commented:
@cormacc , works fine for me. (y)

cormacc commented on layers/+lang/c-c++/README.org:
```diff
@@ -204,3 +297,36 @@ after certain characters by setting the =c-c++-enable-auto-newline= variable.
+
+| Key Binding | Description               |
+|-------------+---------------------------|
+| ~m g &~     | find references (address) |
```
Done

cormacc commented on layers/+lang/c-c++/funcs.el:
```diff
@@ -182,3 +185,134 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+(defun spacemacs//c-c++-lsp-enabled () (member c-c++-backend '(cquery ccls)))
+
+;; -- BEGIN helper functions for common configuration of cquery and ccls backends
+(defun spacemacs//c-c++-lsp-string (prefix suffix) (concat prefix (symbol-name c-c++-backend) suffix))
```
Done

cormacc commented:
Amended the commit incorporating all the requested changes (I think).

@myrgy Your suggestion addressed the rainbow semantic highlighting issue -- I was trying to `apply` a macro as I would a function :S

MaskRay commented on layers/+lang/c-c++/config.el:
```diff
@@ -47,3 +50,33 @@
+
+;; I've left cquery/ccls -extra-init-params separate for now, as one has defaults while the other doesn't
+;; Just to facilitate switching between the two easily
+(defvar c-c++-lsp-extra-init-params '(:cacheFormat "msgpack")
```
Technically ccls does not use "msgpack" but I've just made change to ccls. So as long as the initial is `"m"`, the compact binary serialization format will be used for both ccls and cquery

cormacc replied:
Great - thanks

MaskRay commented on layers/+lang/c-c++/config.el:
```diff
@@ -47,3 +50,33 @@
+initialized. `c-c++-lsp-project-whitelist' is checked first, then this,
+if no pattern matches the project root, lsp-c-c++ will be initialized.")
+
+(defvar c-c++-lsp-sem-highlight-method 'font-lock
```
I would suggest removing this option and refer users to ccls/cquery wiki pages. Some people think semantic highlighting distracting.

MaskRay commented on layers/+lang/c-c++/funcs.el:
```diff
@@ -182,3 +185,167 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+  (spacemacs//c-c++-lsp-call-function "spacemacs//c-c++-lsp-" "-customise-lsp-ui-peek")
+
+  (defun c-c++/callers () (interactive) (lsp-ui-peek-find-custom 'callers (spacemacs//c-c++-lsp-string "$" "/callers")))
+  (defun c-c++/derived () (interactive) (lsp-ui-peek-find-custom 'derived (spacemacs//c-c++-lsp-string "$" "/derived")))
```
`$ccls/derived` `$cquery/derived` have both been removed. Users should instead use lsp-goto-implementation

ccls has a flattened `$ccls/inheritanceHierarchy` to provide multi-level information but I think it is too implementation-specific and does not worth the common configuration. https://github.com/MaskRay/ccls/wiki/Emacs#ccls-cross-reference-extension
```elisp
(defun ccls/derived ()
  (interactive)
  (lsp-ui-peek-find-custom 'derived "$ccls/inheritanceHierarchy"
                           (append (lsp--text-document-position-params) '(:flat t :level 3 :derived t))))
```

MaskRay commented on layers/+lang/c-c++/funcs.el:
```diff
@@ -182,3 +185,167 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+  (defun c-c++/callers () (interactive) (lsp-ui-peek-find-custom 'callers (spacemacs//c-c++-lsp-string "$" "/callers")))
+  (defun c-c++/derived () (interactive) (lsp-ui-peek-find-custom 'derived (spacemacs//c-c++-lsp-string "$" "/derived")))
+  (defun c-c++/vars () (interactive) (lsp-ui-peek-find-custom 'vars (spacemacs//c-c++-lsp-string "$" "/derived")))
+  (defun c-c++/random () (interactive) (lsp-ui-peek-find-custom 'random (spacemacs//c-c++-lsp-string "$" "/derived")))
```
`$ccls/random` `$cquery/random` no longer exist.

sdwolfz commented:
I was thinking of merging #10454 before this layer, hope it won't cause to much trouble for this PR.

cormacc commented:
Rebased
Incorporated feedback from @MaskRay (thanks!).
Corrected a projectile post-config hook that wasn't being applied.
Added `c-c++-adopt-subprojects` layer config variable (defaults to nil) implementing this recommendation from ccls and cquery wikis:
https://github.com/MaskRay/ccls/wiki/Emacs#projects-with-subprojects

@sdwolfz I'll rebase again when you've merged #10454

myrgy commented:
Works fine for me!

cormacc commented:
Thanks @rcoacci - yesterdays amendment resulted in appending a list to projectile-globally-ignored-directories rather than individual strings. Fixed/amended/pushed

rcoacci commented:
Another thing: emacs is hanging _a lot_ when the using the layer, it seems related to the server being busy (indexing probably, `top` shows a lot of cpu being used by `cquery`).  
Also emacs seems to be doing a lot of background work even when its minimized.
It's a very large C++ project, with about 2k source files and lots of libraries.

cormacc commented:
@rcoacci Can't comment there I'm afraid. My use to date has been almost entirely on an embedded C project, which has about 1300 source files, but very few external dependencies. Performance satisfactory here. Some discussion here that may be of some use to you https://github.com/cquery-project/cquery/issues/278

rcoacci commented on layers/+lang/c-c++/README.org:
```diff
@@ -52,22 +66,133 @@ by setting the variable =c-c++-default-mode-for-headers= to =c++-mode=.
+To enable the rainbow semantic highlighting colour theme, set =c-c++-lsp-sem-highlight-rainbow= to =t=.
+
+***** Additional configuration options
+Both supported backends store their index cache in a subdirectory of the project dir by default. This can be overridden by
```
This doesn't seem correct. The default behavior of the layer is to use a subdirectory of spacemacs cache dir as index cache.

cormacc replied:
Good catch - thanks. Changed the default config based on feedback from sdwolfz and forgot to update the README. Corrected / commit amended

Compro-Prasad replied:
I don't think it is the right way to do it. The stuff in .cache is from .emacs.d and not from local project. It is nice to have an option but I don't think it is the right way to do it.

rcoacci commented:
@cormacc I'm willing to blame my project config for now. If I remove the compile_commands.json from the project, it seems to work better. 
I noticed from the logs that it keeps reindexing(?) files with  "Arguments have changed for ..."
I'll test a bit.

MaskRay commented:
@rcoacci Give ccls a try? It leverages some preamble optimization which should accelerate diagnostics,completion a bit. I've also just pushed a commit to optimize indexing when the current file is opened.

kvaneesh commented:
Is there a way to disable lsp-ui-peek features/keybindings and only use xref-* interfaces provided by lsp-mode ? Right now I am adding lsp-ui to dotspacemacs-excluded-packages. But things like xref-find-definitions and all don't have any key binding.

rcoacci commented:
@maskray unfortunately the requirements for ccls are too high for me to build. I'm on a RHEL 7, I'd have to build almost the whole toolchain just to build ccls.

MaskRay commented:
@rcoacci Use cquery then :) But if you'd like to help, it would be good to know if you also come upon https://github.com/MaskRay/ccls/issues/71 

@kvaneesh lsp-ui provides many features. 

> only use xref-* interfaces provided by lsp-mode

If you prefer `ivy-xref` `helm-xref`, you may disable `lsp-ui-peek-mode` by `(setq lsp-ui-peek-enable nil)`

rcoacci commented:
@maskray I'm going on vacation, when I come back I'll take a look. Thanks.

kvaneesh commented:
lsp-ui-peek-enable doesn't look to be used?. However i did this to get it working
```
diff --git a/layers/+tools/lsp/funcs.el b/layers/+tools/lsp/funcs.el
index b6f69eb2fce8..badec151f88d 100644
--- a/layers/+tools/lsp/funcs.el
+++ b/layers/+tools/lsp/funcs.el
@@ -64,13 +64,6 @@ https://github.com/emacs-lsp/lsp-javascript/issues/9#issuecomment-379515379"
     "gt" #'lsp-goto-type-definition
     "gk" #'spacemacs/lsp-avy-document-symbol
     "gm" #'lsp-ui-imenu
-    "gd" #'lsp-ui-peek-find-definitions
-    "gi" #'lsp-ui-peek-find-implementation
-    "gr" #'lsp-ui-peek-find-references
-    "gs" #'lsp-ui-peek-find-workspace-symbol
-    "gp" #'lsp-ui-peek-jump-backward
-    "gn" #'lsp-ui-peek-jump-forward
-    "gf" #'lsp-ui-flycheck-list
     ;;hierarchy
     "hh" #'lsp-describe-thing-at-point
     ;;jump
@@ -87,6 +80,23 @@ https://github.com/emacs-lsp/lsp-javascript/issues/9#issuecomment-379515379"
     "TI" #'spacemacs/lsp-ui-sideline-ignore-duplicate
     )
 
+  (if (not lsp-ui-peek-enable)
+       (spacemacs/set-leader-keys-for-major-mode mode
+        "jd" #'xref-find-definitions
+        "jr" #'xref-find-references
+        "jp" #'xref-pop-marker-stack
+        )
+    (spacemacs/set-leader-keys-for-major-mode mode
+      "jd" #'lsp-ui-peek-find-definitions
+      "ji" #'lsp-ui-peek-find-implementation
+      "jr" #'lsp-ui-peek-find-references
+      "js" #'lsp-ui-peek-find-workspace-symbol
+      "jp" #'lsp-ui-peek-jump-backward
+      "jn" #'lsp-ui-peek-jump-forward
+      "jf" #'lsp-ui-flycheck-list
+      )
+    )
+
   )
```
And have user_config do

```
 (setq lsp-ui-peek-enable nil)
  (setq lsp-ui-sideline-enable nil)
  (setq lsp-ui-doc-enable nil)
```

pitti commented on layers/+lang/c-c++/README.org:
```diff
@@ -52,22 +66,134 @@ by setting the variable =c-c++-default-mode-for-headers= to =c++-mode=.
+  (setq-default dotspacemacs-configuration-layers
+                '((c-c++ :variables
+                         c-c++-backend `lsp-cquery
+                         lsp-c-c++-executable "/path/to/bin/cquery/or/ccls")))
```
I believe this should be `c-c++-lsp-executable`

pitti commented on layers/+lang/c-c++/README.org:
```diff
@@ -52,22 +66,134 @@ by setting the variable =c-c++-default-mode-for-headers= to =c++-mode=.
+  (setq-default dotspacemacs-configuration-layers
+                '((c-c++ :variables
+                         c-c++-backend `lsp-cquery
+                         lsp-c-c++-executable (file-truename "~/bin/cquery/or/ccls"))))
```
This should also be `c-c++-lsp-executable`

cormacc commented:
@pitti Thanks! Corrected the readme and opportunistically rebased on current `develop` branch tip

cormacc commented on layers/+lang/c-c++/README.org:
```diff
@@ -81,7 +193,7 @@ To enable automatic buffer formatting on save, set the variable
 #+BEGIN_SRC emacs-lisp
   (setq-default dotspacemacs-configuration-layers '(
-    (c-c++ :variables c-c++-enable-clang-format-on-save t)))
+                                                    (c-c++ :variables c-c++-enable-clang-format-on-save t)))
```
Moved the `'(` to next line -- means auto-indentation aligns next line under the `d` of `dotspacemacs-configuration-layers`....

MaskRay commented:
Sorry for causing hassle here..

I decide to make requests more orthogonal and thus I have just removed `$ccls/callers`, renamed:

* `$ccls/callHierarchy` -> `$ccls/call`
* `$ccls/inheritanceInheritance` -> `$ccls/inheritance`
* `$ccls/memberHierarchy` -> `$ccls/member`

`c-c++/callers` should be changed to just call `$ccls/call` (it does the same thing as original `$ccls/callers`, though it has `bool hierarchy = false;` which can turn it into a hierarchical view)

`c-c++/call-hierarchy` ... do not need changing as

`ccls-call-hierarchy` `ccls-inheritance-hierarchy` `ccls-member-hierarchy` still exist in emacs-ccls.

MaskRay commented on layers/+lang/c-c++/funcs.el:
```diff
@@ -182,3 +185,161 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+(defun spacemacs//c-c++-lsp-ccls-customise-lsp-ui-peek ()
+  (defun c-c++/base ()
+    (interactive)
+    (lsp-ui-peek-find-custom 'base "$ccls/inheritanceHierarchy"
```
`$ccls/inheritanceHierarchy` -> `$ccls/inheritance`

cormacc replied:
Hey Fangrui,
Will make the changes Friday. A gentle suggestion from a lazy integrator... where it doesn't compromise your vision too much, keeping the ccls API close to that of cquery might encourage uptake, as it makes it easier to support both as drop-in alternatives.

MaskRay commented on layers/+lang/c-c++/funcs.el:
```diff
@@ -182,3 +185,161 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+  "Wrap some backend-specific features using the `lsp-ui-peek' functions provided by the `lsp' layer (`lsp-ui' package)"
+  (spacemacs//c-c++-lsp-call-function "spacemacs//c-c++-lsp-" "-customise-lsp-ui-peek")
+
+  (defun c-c++/callers () (interactive) (lsp-ui-peek-find-custom 'callers (spacemacs//c-c++-lsp-string "$" "/callers")))
```
In ccls, `c-c++/callers` should use 

`(lsp-ui-peek-find-custom 'base "$ccls/call")`

cormacc replied:
Hey Fangrui,
Will make the changes Friday. A gentle suggestion from a lazy integrator... where it doesn't compromise your vision too much, keeping the ccls API close to that of cquery might encourage uptake, as it makes it easier to support both as drop-in alternatives.

MaskRay replied:
Thank you for your kind suggestion (and sorry for the hassle I caused I again). 

The hierarchy APIs are goodies I use a lot but don't expect much adoption by others. ccls is still young and not many people have noticed it. These stuff touched the area where I recently improved (flattened views of old hierarchy APIs) as I know someone favors them (ivy-xref helm-xref integration).

cormacc replied:
No bother Fangrui. I've amended the commit - would you mind verifying the bindings on a known-good project when you get a chance? Both the callers (`, g c`) and base (`, h b`) bindings resulted in a language server timeout for me using ccls, but probably my project setup. Not actively working in C the last couple of weeks, so don't have time to investigate further at the moment.

MaskRay commented:
Also, I have updated `lsp-ui-peek-find-custom` to match that of `lsp-find-custom`

`(lsp--text-document-position-params)` is always passed in the params.

As what I described in ccls/wiki/Emacs

```
(lsp-ui-peek-find-custom 'caller "$ccls/call")
(lsp-ui-peek-find-custom 'callee "$ccls/call" '(:callee t))

;; If you prefer ivy-xref helm-xref
(lsp-find-custom "$ccls/call")
(lsp-find-custom "$ccls/call" '(:callee t))
```

cquery's requests should be similar.

MaskRay commented on layers/+lang/c-c++/funcs.el:
```diff
@@ -182,3 +185,161 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+  (defun c-c++/base ()
+    (interactive)
+    (lsp-ui-peek-find-custom 'base "$ccls/inheritanceHierarchy"
+      (append (lsp--text-document-position-params) '(:flat t :level 3)))))
```
`:level 3` -> `:levels 1` for direct bases, `:levels 3` for bases up to 3 levels.

MaskRay commented on layers/+lang/c-c++/funcs.el:
```diff
@@ -182,3 +185,161 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+  (defun c-c++/call-hierarchy-inv () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-call-hierarchy" t))
+  (defun c-c++/inheritance-hierarchy () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-inheritance-hierarchy"))
+  (defun c-c++/inheritance-hierarchy-inv () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-inheritance-hierarchy" t))
+  (defun c-c++/member-hierarchy () (interactive) (spacemacs//c-c++-lsp-funcall-interactively-no-args nil "-member-hierarchy"))
```
memberHierarchy has been deleted from cquery.

cormacc replied:
:) Sounds like it may be reinstated:
https://github.com/cquery-project/cquery/issues/789
I'll hang tough for a few days and see what happens

cormacc replied:
... and it seems to have been reinstated

MaskRay commented on layers/+lang/c-c++/funcs.el:
```diff
@@ -182,3 +185,161 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+  (defun c-c++/inheritance-hierarchy () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-inheritance-hierarchy"))
+  (defun c-c++/inheritance-hierarchy-inv () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-inheritance-hierarchy" t))
+  (defun c-c++/member-hierarchy () (interactive) (spacemacs//c-c++-lsp-funcall-interactively-no-args nil "-member-hierarchy"))
+  (defun c-c++/freshen-index () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-freshen-index"))
```
`$cquery/freshenIndex` `$ccls/reload`... these might be to specific and should not be included.

cormacc replied:
Useful in some circumstances and bound under `SPC m l` (i.e. not cluttering up the 'goto' bindings or anything). I'm inclined to leave them...

myrgy replied:
@cormacc , seems like `ccls-reload` call was missed. It's very useful call when for example git branch was changed.

cormacc replied:
Well spotted myrgy -- corrected. Also moved those bindings from 'l' to 'b' prefix to reflect change in mnemonic in lsp-layer.

MaskRay commented on layers/+lang/c-c++/funcs.el:
```diff
@@ -182,3 +185,161 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+    "gf" 'find-file-at-point
+    "gF" 'ffap-other-window
+    "g&" #'c-c++/references-address
+    "gR" #'c-c++/references-read
```
I also don't know if `read/write/address` references should be included as default key bindings.

MaskRay commented on layers/+lang/c-c++/funcs.el:
```diff
@@ -182,3 +185,162 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+  (defun c-c++/base ()
+    (interactive)
+    (lsp-ui-peek-find-custom 'base "$ccls/inheritance"
+      (append (lsp--text-document-position-params) '(:flat t :level 3)))))
```
`c-c++/base` for ccls can be:

```elisp
(lsp-ui-peek-find-custom 'base "$ccls/inheritance" '(:levels 3))
```

`:levels 3` is for bases up to 3 levels above. It can be removed (`levels` defaults to 1) if direct bases are needed.

cormacc replied:
Thanks fangrui -- I cleverly hadn't thought to recompile ccls in a while :)

MaskRay replied:
Since `$ccls/memberHierarchy` has been renamed to ``$ccls/member`, it makes me feel more comfortable to add other non hierarchy features, e.g. listing member functions of a class or listing nested classes/namespace members

```
(lsp-find-custom "$ccls/member" `(:kind 2)) => nested classes / types in a namespace
(lsp-find-custom "$ccls/member" `(:kind 3)) => member functions / functions in a namespace
(lsp-find-custom "$ccls/member" `(:kind 0)) => other than 2/3 => member variables / variables in a namespace
```

Diagnostics have also been improved much and memory footprint has been improved a lot.

cormacc replied:
Thanks Fangrui. I've added bindings for them under `SPC m <g/p> m <c/f/v>` when ccls is selected as backend. Won't amend/push until #11351 gets merged/rejected, as I'm using those lsp-layer changes in my local c-c++ layer.

cormacc commented:
@kvaneesh Your comments relate to the lsp layer rather than this PR, but echo sentiments expressed by a few others in the recent past. I've raised a new PR against the lsp layer that should hopefully address them: #11351 
@sdwolfz You might take a look at that new PR when you get a chance? If that gets merged I'll rebase this one, and may make some minor related modifications.

MaskRay commented on layers/+lang/c-c++/packages.el:
```diff
@@ -254,3 +260,46 @@
+        (add-to-list 'projectile-globally-ignored-directories c-c++-lsp-cache-dir))
+      (when c-c++-adopt-subprojects
+        (setq projectile-project-root-files-top-down-recurring
+          (append '("compile_commands.json"
```
For the root indicators, you may want to join discussions here (beyond this PR)

https://github.com/emacs-lsp/lsp-mode/issues/293#issuecomment-423728467

myrgy commented:
Agree with @MaskRay , because I use this MR on daylight basis using.
There were a few issues with ccls, need to check my local patches.

cormacc commented:
@myrgy Pretty sure those ccls issues are fixed in my local version, but that's based on #11351 , so deferring push until that's merged.

MaskRay commented:
Ping

sergei-dyshel commented:
Hi, I don't know if it's the proper place to ask my question (if not - please suggest me) but I'm looking for a way to use `ycmd` for completion and diagnostics and use c++ LSP (ccls/cquery) for everything else. Also I want this to be configurable per project (projectile's): for some projects I want to use ycmd/LSP, and for others I want to use `ccls` alone (with completion provided by lsp) and for others I want to use `cquery` alone. Can you help me with the needed elisp (at least hints)?

I've just tried Spacemacs yesterday and currently I'm basing on this PR's commmit.
Thanks!

d12frosted commented:
Wow. What a PR. 

@sdwolfz what do you think about it?

sdwolfz commented:
I'm trying to find the time to review this properly and merge, but I'm optimistic about it.

cormacc commented:
@sdwolfz I've rebased the blocking #11351 on develop if you want to take a look at that first. Think it should be a pretty safe merge as the changes are in default core keybinding functions I contributed myself and no other layers are using. I'll rebase and push a new commit of this PR once that one gets resolved either way.

iromise commented on layers/+lang/c-c++/README.org:
```diff
@@ -52,22 +66,134 @@ by setting the variable =c-c++-default-mode-for-headers= to =c++-mode=.
+
+***** Additional configuration options
+Both lsp backends are configured to store their index cache in a subdirectory of =.emacs.d/cache=. This can be overridden by
+specifying an explicit =lsp-c-c++-cache-dir=. Setting this value to a relative path will cause the index cache to be placed in a
```
Hi, `lsp-c-c++-cache-dir` should be `c-c++-lsp-cache-dir` because 

```lisp
;; c-c++-lsp-backend variables
(defvar c-c++-lsp-cache-dir nil
  "Cache directory. Absolute and relative paths supported.")
```

cormacc replied:
Thanks iromise -- corrected

cormacc commented:
@myrgy I've pushed an updated version of this based on #11351 that I've been sitting on for a while - you might let me know if it resolves the ccls issues you were patching around locally if you get a chance to evaluate? I haven't had any issues in the last few weeks, though admittedly have been doing more python and planning/scoping than C in that period...

myrgy commented:
@cormacc , thank you! I'll give a try on thursday

cormacc commented on layers/+lang/c-c++/README.org:
```diff
@@ -52,22 +66,134 @@ by setting the variable =c-c++-default-mode-for-headers= to =c++-mode=.
+  (setq-default dotspacemacs-configuration-layers
+                '((c-c++ :variables
+                         c-c++-adopt-subprojects t
+                         c-c++-backend 'ccls
```
Thanks svenihoney - updated the readme. @sdwolfz , you happy enough to use the 'Apply suggestion' feature as a general rule? Looks very convenient for this sort of change (oversights in the README). I'd imagine it probably results in a new commit though rather than an amendment (which appears to be the recommended approach in the beginner's guide to PRs)? Regardless, went with the traditional `git commit --amend`, `git push --force` in this instance, pending enlightenment :)

sdwolfz replied:
It's a nice feature but I would like to have as few commits as possible at the end. So if they are small fixups then I'd preffer they be squashed anyway.

cormacc commented:
@sdwolfz Rebased / pushed after merge of #11351

myrgy commented:
@cormacc , I gave a try - seems to work fine for me :+1:. (ccls)

sdwolfz commented:
@JAremko the `Validate Documentation (required)` job is failing for this build and I have no idea why, can you please take a look?

myrgy commented:
Could it be the reason?
```
=============== WARNING ===============
Missing "#+TAGS:" keyword. See "https://github.com/syl20bnr/spacemacs/blob/develop/core/templates/README.org.template"
=======================================
```

JAremko commented on layers/+lang/c-c++/README.org:
```diff
@@ -52,22 +66,134 @@ by setting the variable =c-c++-default-mode-for-headers= to =c++-mode=.
+                '((c-c++ :variables c-c++-backend `lsp-cquery))) ;or 'lsp-ccls
+#+END_SRC
+
+N.B. The [[../../+tools/lsp/README.org][LSP layer]] will be loaded automatically if either backend is selected.
```
@sdwolfz  Should be `[[file:../../+tools/lsp/README.org][LSP layer]] `  
Actually the current one is also valid but we a bit strict about it. Mb too strict  :thinking:

JAremko replied:
I'll add it to autofixes.

sdwolfz commented:
Thank you :heart:!

There was a problem with the `c-c++-backend` variable when it was unbound, an error was thrown since the `layers.el` file is loaded before the `config.el` file, but I addressed that and amended it to the commit: 445f6af93f6d5555bbda7608713bd4838c9b8274.

Cherry-picked into develop branch, you can safely delete your branch.

MaskRay commented:
@sdwolfz Please also take a look at https://github.com/syl20bnr/spacemacs/pull/11583

I didn't mention it as this was already taking too long.

vityafx commented:
Sorry, I am kinda new to all of this LSP thing. How can I setup it easily for my spacemacs?

Compro-Prasad commented:
@vityafx Please open an issue if you find something missing or open a PR if you have a solution or discuss about this [somewhere else](https://gitter.im/syl20bnr/spacemacs).
